modules/gmc-agent/src/main/java/org/gridgain/action/controller/QueryActionsController.java

```diff
@@ -0,0 +1,188 @@
+
+    /**
+     * @param ctx Context.
+     */
+    public QueryActionsController(GridKernalContext ctx) {
+        this.ctx = ctx;
+        this.log = ctx.log(QueryActionsController.class);
+
+        qryProc = ctx.query();
+        qryRegistry = new QueryHolderRegistry(ctx, Duration.ofMinutes(5));
+    }
+
+    /**
+     * Cancel query by query ID.
+     *
+     * @param qryId Query id.
+     */
+    public CompletableFuture<Void> cancel(String qryId) {
+        qryRegistry.cancelQuery(qryId);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    /**
+     * @param arg Argument.
+     * @return Next page with result.
+     */
+    public CompletableFuture<QueryResult> nextPage(NextPageQueryArgument arg) {
+        final CompletableFuture<QueryResult> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            try {
+                String qryId = arg.getQueryId();
+                if (F.isEmpty(qryId))
+                    throw new IllegalArgumentException("Fail to execute query - query id can't be empty.");
+
+                String cursorId = arg.getCursorId();
+                if (F.isEmpty(cursorId))
+                    throw new IllegalArgumentException("Fail to execute query - cursor id can't be empty.");
+
+                CursorHolder cursorHolder = qryRegistry.findCursor(qryId, cursorId);
+                QueryResult res = fetchResult(qryId, cursorHolder, arg.getPageSize());
+
+                if (!cursorHolder.hasNext())
+                    qryRegistry.closeQueryCursor(qryId, cursorId);
+
+                fut.complete(res);
+            }
+            catch (Throwable e) {
+                fut.completeExceptionally(e);
+            }
+        }, MANAGEMENT_POOL);
+
+        return fut;
+    }
+
+    /**
+     * @param arg Argument.
+     * @return List of query results.
+     */
+    public CompletableFuture<List<QueryResult>> executeSqlQuery(QueryArgument arg) {
+        final CompletableFuture<List<QueryResult>> fut = new CompletableFuture<>();
+
+        ctx.closure().runLocalSafe(() -> {
+            String qryId = arg.getQueryId();
+            qryRegistry.cancelQuery(qryId);
+
+            QueryHolder qryHolder = qryRegistry.createQueryHolder(qryId);
+
+            try {
+                if (log.isDebugEnabled())
+                    log.debug("Operation started with subject: " + ctx.security().securityContext().subject());
+
+                SqlFieldsQuery qry = prepareQuery(arg);
+                GridCacheContext cctx = !F.isEmpty(arg.getCacheName())
+                        ? ctx.cache().cache(arg.getCacheName()).context()
+                        : null;
+
+                List<QueryResult> results = new ArrayList<>();
+                for (FieldsQueryCursor<List<?>> cur : qryProc.querySqlFields(cctx, qry, null, true, false, qryHolder.getCancelHook()))
+                    results.add(fetchResult(qryId, new CursorHolder(cur), arg.getPageSize()));
```

Comment:
I think it's ok all SQL tools don't run next queries if previous query fails. I've removed exception if we don't have meta information, instead of it we return empty list.

Reply:
You already executed all queries and ignore results if reading result of one failed.